### PR TITLE
feat: Make shallowRender configurable

### PR DESCRIPTION
**Pull request by necatiozmen**
We wanted to make optional SSR when trigger changeLanguage method. So, we set ` shallowRender: false ` as a default value in `default-config.ts`. 

After this, users can set `shallowRender: true` in order to disable SSR on each changeLanguage method trigger.

### CODE REVIEW

**Comment by isaachinman** [comment]
Please clarify. If user action results in `changeLanguage` being called, it's on the client side and has nothing to do with SSR.

**Comment by necatiozmen** [comment]
Correct, With disable SSR, I meant disabling triggering getInitialProps

**Comment by isaachinman** [comment]
Needs a semantic title in order to pass all checks and be ready for review.

**Comment by necatiozmen** [comment]
Done @isaachinman

File: src/hocs/app-with-translation.tsx [hunk 1]
```diff
@@ -43,7 +43,7 @@ export default function (WrappedComponent) {
 
           if (i18n.initializedLanguageOnce && typeof newLng === 'string' && prevLng !== newLng) {
             const { as, href } = lngPathCorrector(config, { as: asPath, href: routeInfo }, newLng)
-            router.replace(href, as)
+            router.replace(href, as, {shallow: config.shallowRender})
```
Comment (isaachinman) [inline src/hocs/app-with-translation.tsx]: Please reformat to include spaces around the curly brackets.

**Comment by necatiozmen** [comment]
Added test coverage and linting update

**Comment by isaachinman** [comment]
Hi @necatiozmen - you've added test coverage for truthiness of a boolean. I was referring to coverage that asserts whether or not `getInitialProps` is called. Do you think that'd be possible?

**Comment by necatiozmen** [comment]
Hi @isaachinman,
Since getInitialProps is a  Next.js property, I think it should be tested in Next.js.

**Comment by isaachinman** [comment]
We're not testing the function, we're testing whether or not it gets called.

**Comment by schneck** [comment]
@isaachinman I'd agree with @necatiozmen's argument that `getInitialProps` not being called is just a side effect of shallow routing on route change, and there's no need to test this explicitely. However, it might by an idea to test whether the route change was actually performed using `{ shallow: true }`, like:

```
describe('appWithTranslation', () => {
  […]
  it('will call change language with shallow routing if shallowRender is true', async () => {
    isServer.mockReturnValue(false)
    const { i18n } = await createApp({...defaultConfig, shallowRender: true});
    (i18n as any).initializedLanguageOnce = true
    await i18n.changeLanguage('de')
    expect(mockRouterFn).toHaveBeenCalledWith(
      {"pathname": defaultProps.router.pathname, "query": {}}, 
      defaultProps.router.pathname, 
      { shallow: true}
    )
  })

  it('will not call change language with shallow routing if shallowRender is false', async () => {
    isServer.mockReturnValue(false)
    const { i18n } = await createApp();
    (i18n as any).initializedLanguageOnce = true
    await i18n.changeLanguage('de')
    expect(mockRouterFn).toHaveBeenCalledWith(
      {"pathname": defaultProps.router.pathname, "query": {}}, 
      defaultProps.router.pathname, 
      { shallow: false}
    )
  })
```

**Comment by necatiozmen** [comment]
Hi @schneck, thanks for the sharing. I can add the @schneck 's test code  if you it's ok for you @isaachinman

**Comment by isaachinman** [comment]
Yup, that'll do!

**Comment by necatiozmen** [comment]
I added test but can not push the commit due to errors. 

Screen Shot 2019-11-27 at 15 38 22

**Comment by isaachinman** [comment]
@necatiozmen Is that the same issue as #574?

**Comment by isaachinman** [comment]
@necatiozmen That issue is now resolved. What's the latest here?

**Comment by necatiozmen** [comment]
@isaachinman I pulled the latest update but there are still errors.

Screen Shot 2019-12-04 at 11 27 54

**Comment by isaachinman** [comment]
@hgezim Any idea?

**Comment by isaachinman** [comment]
@necatiozmen What version of TypeScript are you using? I can't understand if that error is complaining about duplicate instantiation due to case insensitivity, or if it's literally saying that `i18n` within types/interfaces must be identical. Can you give some more context?

**Comment by necatiozmen** [comment]
@isaachinman  I use 3.7.2 version of tsc. Do you have any suggestion ?  I can disable husky and push the code so you can see the errors on your local if you want.

**Comment by isaachinman** [comment]
Yes, please push with `--no-verify`.

**Comment by isaachinman** [comment]
> I use 3.7.2 version of tsc

Do you mean you have a different TypeScript version installed globally on your machine? It looks like this PR compiles fine and passes all tests. What's the issue? Even if I upgrade your fork to `typescript@3.7.3`, it runs fine.

Did you mean that you are seeing an issue in TypeScript consumers of `next-i18next`?

**Comment by Accumulative** [comment]
The typescript check runs fine for me. I don't see anything stopping this being merged in already but I can created another pull request with the conflicts resolved if desired?

**Comment by isaachinman** [comment]
@Accumulative Yeah, I couldn't reproduce it either. Yes, I'll close this PR now due to inactivity, and please do open another with resolved conflicts.

**Comment by Pierre-CLIND** [comment]
Thanks a lot, it's working ;)

**Pull request closed without merging**